src/main/java/tc/oc/pgm/rotation/VotingPool.java

```diff
@@ -0,0 +1,158 @@
+package tc.oc.pgm.rotation;
+
+import app.ashcon.intake.CommandException;
+import java.util.*;
+import net.md_5.bungee.api.ChatColor;
+import net.md_5.bungee.api.chat.ClickEvent;
+import net.md_5.bungee.api.chat.HoverEvent;
+import org.bukkit.configuration.ConfigurationSection;
+import tc.oc.component.Component;
+import tc.oc.component.types.PersonalizedText;
+import tc.oc.component.types.PersonalizedTranslatable;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.map.PGMMap;
+
+public class VotingPool extends MapPool {
+
+  // Number of maps in the vote, unless not enough maps in pool
+  private static final int MAX_VOTE_OPTIONS = 5;
+
+  private final double DEFAULT_WEIGHT;
+  private final int VOTE_SIZE;
+  private final Map<PGMMap, Double> mapScores = new HashMap<>();
+
+  private Poll currentPoll;
+
+  public VotingPool(MapPoolManager manager, ConfigurationSection section, String name) {
+    super(manager, section, name);
+    DEFAULT_WEIGHT = 1d / maps.size();
+    VOTE_SIZE = Math.min(MAX_VOTE_OPTIONS, maps.size());
+
+    for (PGMMap map : maps) {
+      mapScores.put(map, DEFAULT_WEIGHT);
+    }
+  }
+
+  public Poll getCurrentPoll() {
+    return currentPoll;
+  }
+
+  /** Ticks scores for all maps, making them go slowly towards DEFAULT_WEIGHT. */
+  private void tickScores() {
+    mapScores.replaceAll((mapScores, value) -> (value * 3 + DEFAULT_WEIGHT) / 4);
+  }
+
+  @Override
+  public PGMMap popNextMap() {
+    if (currentPoll == null) return getRandom();
+
+    PGMMap map = currentPoll.finishVote();
+    currentPoll = null;
+    tickScores();
+    return map;
+  }
+
+  @Override
+  public PGMMap getNextMap() {
+    return currentPoll == null ? null : currentPoll.getMostVotedMap();
+  }
+
+  @Override
+  public void setNextMap(PGMMap map) {
+    currentPoll = null;
+  }
+
+  @Override
+  public void matchEnded(Match match) {
+    if (manager.getOverriderMap() != null) return;
+    currentPoll = new Poll();
+    match.getPlayers().forEach(currentPoll::sendMessage);
+  }
+
+  /** Represents a polling process, with a set of options. */
+  public class Poll {
```

Comment:
If we're exposing this, how about we make it a separate class?

Reply:
Can't really make it a separate class, it's not a static inner class, it takes info from the VotingPool directly. Also it's not a class that i think will be reused, so it's fine if it stays there.